pkg/utils/symbol_table.go

```diff
@@ -0,0 +1,212 @@
+type SymbolTable[T Symbol] struct {
+	mu sync.RWMutex
+	// All symbols sorted by their address in descending order,
+	// for quick binary searches by address.
+	sortedSymbols []*T
+	// If lazyNameLookup is true, the symbolsByName map
+	// will be populated only when a failed lookup occurs.
+	symbolsByName  map[string][]*T
+	lazyNameLookup bool
+}
+
+var ErrSymbolNotFound = errors.New("symbol not found")
+
+// Creates a new SymbolTable. If lazyNameLookup is true, the mapping from
+// name to symbol will be populated only when a failed lookup occurs.
+// This reduces memory footprint at the cost of the time it takes to lookup
+// a symbol name for the first time.
+func NewSymbolTable[T Symbol](lazyNameLookup bool) *SymbolTable[T] {
+	return &SymbolTable[T]{
+		sortedSymbols:  make([]*T, 0),
+		symbolsByName:  make(map[string][]*T),
+		lazyNameLookup: lazyNameLookup,
+	}
+}
+
+// Adds a slice of symbols to the symbol table.
+func (st *SymbolTable[T]) AddSymbols(symbols []*T) {
+	st.mu.Lock()
+	defer st.mu.Unlock()
+
+	// Add the new symbols to the sorted slice (which now becomes unsorted).
+	// Allocate the slice with the needed capacity to avoid overallocation.
+	oldSymbols := st.sortedSymbols
+	newLen := len(oldSymbols) + len(symbols)
+	st.sortedSymbols = make([]*T, 0, newLen)
+	st.sortedSymbols = append(st.sortedSymbols, oldSymbols...)
+
+	// If lazyNameLookup is false, we update the name to symbol mapping for
+	// each new symbol
+	if !st.lazyNameLookup {
+		for _, symbol := range symbols {
+			st.sortedSymbols = append(st.sortedSymbols, symbol)
+			name := (*symbol).Name()
+			if symbols, found := st.symbolsByName[name]; found {
+				st.symbolsByName[name] = append(symbols, symbol)
+			} else {
+				st.symbolsByName[name] = []*T{symbol}
+			}
+		}
+	} else {
+		st.sortedSymbols = append(st.sortedSymbols, symbols...)
+	}
+
+	// Sort the symbols slice by address in descending order
+	sort.Slice(st.sortedSymbols,
+		func(i, j int) bool {
+			return (*st.sortedSymbols[i]).Address() > (*st.sortedSymbols[j]).Address()
+		})
+}
+
+// Lookup a symbol in the table by its name.
+// Because there may be multiple symbols with the same name, a slice of all
+// matching symbols is returned.
+func (st *SymbolTable[T]) LookupByName(name string) ([]*T, error) {
+	st.mu.RLock()
+	// We call RUnlock manually and not using defer because we may need to upgrade to a write lock later
+
+	// Lookup the name in the name to symbol mapping
+	if symbols, found := st.symbolsByName[name]; found {
+		st.mu.RUnlock()
+		return symbols, nil
+	}
+
+	// Lazy name lookup is disabled, the lookup failed
+	if !st.lazyNameLookup {
+		st.mu.RUnlock()
+		return nil, ErrSymbolNotFound
+	}
+
+	// Lazy name lookup is enabled, perform a linear search to find the requested name
+	symbols := []*T{}
+	for _, symbol := range st.sortedSymbols {
+		if (*symbol).Name() == name {
+			symbols = append(symbols, symbol)
+		}
+	}
+	st.mu.RUnlock()
```

Comment:
You should for safety's sake push one of these into the start of the following `if` block, and then duplicate this call before the last `return`.

Reply:
Not sure I understand what you mean, why is that safer?

Reply:
I assume you're using a mutex because different goroutines can call this data structure's methods from multiple locations. By unlocking here you're leaving a small gap where another goroutine can "sneak in", do changes and then unlock you back inside the `if` post changes. 
That's why the gap between locks and unlocks should be as small as possible - so this unlock should be done near the inner `Lock` of the `if` or right before the return.

Reply:
Understood :thumbsup: